src/Microsoft.AspNetCore.Server.Kestrel/Http/Listener.cs

```diff
@@ -84,6 +84,8 @@ protected virtual void DispatchConnection(UvStreamHandle socket)
 
         public virtual async Task DisposeAsync()
         {
+            Memory2.Dispose();
```

Comment:
I makes sense that calling `Memory2.Dispose()` here will prevent the assert failures, but so would removing the assert :stuck_out_tongue:

Joking aside, this *is* better than removing the assert since the assert can still catch non-shutdown related memory leaks which are likely the most important. However, since the assert only throws for dev builds and most of our tests shutdown the server fairly quickly, even non-shutdown related leaks might not be caught.

The reason Memory2.Dispose is called after shutdown is to suppress finallization of the existing blocks, and prevent new slabs from being allocated for blocks leased during shutdown.

I would rather find out why DisposeAsync isn't completing within the timeout. If we can't the more targeted fix would be to dispose the pools [here in KestrelEngine only if graceful shutdown fails](https://github.com/aspnet/KestrelHttpServer/blob/bb2e76c7f1a23fe4fbe0a97b29a04bfe06b7efde/src/Microsoft.AspNetCore.Server.Kestrel/KestrelEngine.cs#L119).

But to reiterate, I would *much* rather find out what's going on with DisposeAsync timing out. If the `PostAsync` callback isn't getting invoked, I really want to know why.